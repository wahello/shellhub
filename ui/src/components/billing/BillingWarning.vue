<template>
  <fragment>
    <v-dialog
      v-model="dialog"
      max-width="510"
    >
      <v-card data-test="billingWarning-dialog">
        <v-card-title class="headline grey lighten-2 text-center">
          Update account
        </v-card-title>

        <v-card-text class="mt-4 mb-3 pb-1">
          <p>
            This namespace has maximum number of devices on your free account.
          </p>

          <p>
            If you create the subscription in your account settings, you can continue to take
            advantage of the features available on ShellHub by adding more devices.
          </p>
        </v-card-text>

        <v-card-actions>
          <v-spacer />

          <v-btn
            text
            data-test="close-btn"
            @click="dialog=!dialog"
          >
            Close
          </v-btn>

          <v-btn
            to="/settings/billing"
            text
            data-test="goToBilling-btn"
            @click="dialog=!dialog"
          >
            Go to Billing
          </v-btn>
        </v-card-actions>
      </v-card>
    </v-dialog>
  </fragment>
</template>

<script>

export default {
  name: 'BillingWarning',

  data() {
    return {
      dialog: true,
    };
  },

  computed: {
    active() {
      return this.$store.getters['billing/active'];
    },

    stats() {
      return this.$store.getters['stats/stats'];
    },
  },

  created() {
    this.dialog = this.stats.registered_devices === 3 && !this.active;
  },
};

</script>
