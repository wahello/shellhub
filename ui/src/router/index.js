import Vue from 'vue';
import Router from 'vue-router';
import Dashboard from '@/views/Dashboard';
import store from '@/store';

Vue.use(Router);

const router = new Router({
  mode: 'history',
  base: process.env.BASE_URL,
  routes: [
    {
      path: '/',
      name: 'dashboard',
      component: Dashboard,
    },
    {
      path: '/sign-up',
      name: 'signUp',
      component: () => import('@/views/SignUp'),
    },
    {
      path: '/validation-account',
      name: 'validationAccount',
      component: () => import('@/views/ValidationAccount'),
    },
    {
      path: '/forgot-password',
      name: 'forgotPassword',
      component: () => import('@/views/ForgotPassword'),
    },
    {
      path: '/update-password',
      name: 'updatePassword',
      component: () => import('@/views/UpdatePassword'),
    },
    {
      path: '/devices',
      name: 'devices',
      component: () => import(/* webpackChunkName: 'devices' */ '@/views/Devices'),
      redirect: {
        name: 'listDevices',
      },
      children: [
        {
          path: '',
          name: 'listDevices',
          component: () => import('@/components/device/DeviceList'),
        },
        {
          path: 'pending',
          name: 'pendingDevices',
          component: () => import('@/components/device/DevicePendingList'),
        },
        {
          path: 'rejected',
          name: 'rejectedDevices',
          component: () => import('@/components/device/DeviceRejectedList'),
        },
      ],
    },
    {
      path: '/device/:id',
      name: 'detailsDevice',
      component: () => import(/* webpackChunkName: 'details-device' */ '@/views/DetailsDevice'),
    },
    {
      path: '/sessions',
      name: 'sessions',
      component: () => import(/* webpackChunkName: 'sessions' */ '@/views/Sessions'),
      redirect: {
        name: 'listSessions',
      },
      children: [
        {
          path: '',
          name: 'listSessions',
          component: () => import('@/components/session/SessionList'),
        },
      ],
    },
    {
      path: '/session/:id',
      name: 'detailsSession',
      component: () => import(/* webpackChunkName: 'details-session' */ '@/views/DetailsSession'),
    },
    {
      path: '/login',
      name: 'login',
      component: () => import('@/views/Login'),
    },
    {
      path: '/firewall/rules',
      name: 'firewalls',
      component: () => import(/* webpackChunkName: 'firewalls' */ '@/views/FirewallRules'),
      redirect: {
        name: 'listFirewalls',
      },
      children: [
        {
          path: '',
          name: 'listFirewalls',
          component: () => import('@/components/firewall_rule/FirewallRuleList'),
        },
      ],
    },
    {
      path: '/sshkeys/public-keys',
      name: 'publicKeys',
      component: () => import(/* webpackChunkName: 'publickeys' */'@/views/PublicKeys'),
      redirect: {
        name: 'listPublickeys',
      },
      children: [
        {
          path: '',
          name: 'listPublickeys',
          component: () => import('@/components/public_key/PublicKeyList'),
        },
      ],
    },
    {
      path: '/settings',
      name: 'settings',
      component: () => import(/* webpackChunkName: 'settings' */ '@/views/Settings'),
      redirect: {
        name: 'profileSettings',
      },
      children: [
        {
          path: 'profile',
          name: 'profileSettings',
          component: () => import('@/components/setting/SettingProfile'),
        },
        {
          path: 'namespace-manager',
          name: 'namespaceSettings',
          component: () => import('@/components/setting/SettingNamespace'),
        },
        {
          path: 'private-keys',
          name: 'privateKeysSettings',
          component: () => import('@/components/setting/SettingPrivateKeys'),
        },
        {
          path: 'tags',
          name: 'tagsSettings',
          component: () => import('@/components/setting/SettingTags'),
        },
        {
          path: 'billing',
          name: 'billingSettings',
          component: () => import('@/components/setting/SettingBilling'),
        },
      ],
    },
    {
      path: '*',
      name: 'NotFound',
      component: Dashboard,
      redirect: () => {
        localStorage.setItem('flag', true);
        return '/';
      },
    },
  ],
});

router.beforeEach((to, from, next) => {
  if ((to.path !== '/login' && to.path !== '/sign-up') && to.path !== '/forgot-password' && to.path !== '/validation-account' && to.path !== '/update-password') {
    if (store.getters['auth/isLoggedIn']) {
      return next();
    }
    return next(`/login?redirect=${to.path}`);
  }
  if (store.getters['auth/isLoggedIn']) {
    if (to.path === '/login' && to.query.token) {
      return next();
    }
    return next('/');
  }
  if (to.path === '/sign-up' && process.env.VUE_APP_SHELLHUB_CLOUD === 'false') {
    return next('/');
  }
  return next();
});

export default router;
