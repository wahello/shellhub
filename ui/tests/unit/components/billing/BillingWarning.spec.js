import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import Vuetify from 'vuetify';
import BillingWarning from '@/components/billing/BillingWarning';
import router from '@/router/index';

describe('BillingWarning', () => {
  const localVue = createLocalVue();
  const vuetify = new Vuetify();
  localVue.use(Vuex);

  document.body.setAttribute('data-app', true);

  let wrapper;

  let stateBilling = false;

  const stats = {
    registered_devices: 0,
    online_devices: 0,
    active_sessions: 0,
    pending_devices: 1,
    rejected_devices: 0,
  };

  const storeHasZeroDevices = new Vuex.Store({
    namespaced: true,
    state: {
      stateBilling,
      stats,
    },
    getters: {
      'billing/active': (state) => state.stateBilling,
      'stats/stats': (state) => state.stats,
    },
    actions: {
    },
  });

  const storeHasThreeDevices = new Vuex.Store({
    namespaced: true,
    state: {
      stateBilling,
      stats: { ...stats, registered_devices: 3 },
    },
    getters: {
      'billing/active': (state) => state.stateBilling,
      'stats/stats': (state) => state.stats,
    },
    actions: {
    },
  });

  const storeHasThreeDevicesWithSubscription = new Vuex.Store({
    namespaced: true,
    state: {
      stateBilling: true,
      stats: { ...stats, registered_devices: 3 },
    },
    getters: {
      'billing/active': (state) => state.stateBilling,
      'stats/stats': (state) => state.stats,
    },
    actions: {
    },
  });

  ///////
  // In this case, the test dialog is closes when the user has less
  // than 3 devices and has no subscription.
  ///////

  describe('Dialog is closes when has less than 3 devices', () => {
    beforeEach(() => {
      wrapper = mount(BillingWarning, {
        localVue,
        store: storeHasZeroDevices,
        stubs: ['fragment'],
        router,
        vuetify,
      });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    ///////

    it('Compare data with default value', () => {
      expect(wrapper.vm.dialog).toEqual(false);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with data', async () => {
      expect(wrapper.find('[data-test="billingWarning-dialog"]').exists()).toBe(false);
      expect(wrapper.find('[data-test="close-btn"]').exists()).toBe(false);
      expect(wrapper.find('[data-test="goToBilling-btn"]').exists()).toBe(false);
    });
  });

  ///////
  // In this case, the test dialog is closes when the user has 3
  // devices and has subscription.
  ///////

  describe('Dialog is closes when user has subscription', () => {
    beforeEach(() => {
      wrapper = mount(BillingWarning, {
        localVue,
        store: storeHasThreeDevicesWithSubscription,
        stubs: ['fragment'],
        router,
        vuetify,
      });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    ///////

    it('Compare data with default value', () => {
      expect(wrapper.vm.dialog).toEqual(false);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with data', async () => {
      expect(wrapper.find('[data-test="billingWarning-dialog"]').exists()).toBe(false);
      expect(wrapper.find('[data-test="close-btn"]').exists()).toBe(false);
      expect(wrapper.find('[data-test="goToBilling-btn"]').exists()).toBe(false);
    });
  });

  ///////
  // In this case, the test dialog is opened when the user has 3
  // devices and has no subscription.
  ///////

  describe('Dialog is opens', () => {
    stateBilling = false;

    beforeEach(() => {
      wrapper = mount(BillingWarning, {
        localVue,
        store: storeHasThreeDevices,
        stubs: ['fragment'],
        router,
        vuetify,
      });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    ///////

    it('Compare data with default value', () => {
      expect(wrapper.vm.dialog).toEqual(true);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with data', async () => {
      expect(wrapper.find('[data-test="billingWarning-dialog"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="close-btn"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="goToBilling-btn"]').exists()).toBe(true);
    });
  });
});
