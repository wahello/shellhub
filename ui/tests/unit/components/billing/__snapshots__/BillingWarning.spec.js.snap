// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BillingWarning Dialog is closes when has less than 3 devices Renders the component 1`] = `
<fragment-stub>
  <div role="dialog" class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;

exports[`BillingWarning Dialog is closes when user has subscription Renders the component 1`] = `
<fragment-stub>
  <div role="dialog" class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;

exports[`BillingWarning Dialog is opens Renders the component 1`] = `
<fragment-stub>
  <div role="dialog" class="v-dialog__container"></div>
</fragment-stub>
`;
